ocaml/tests/test_observer.ml

```diff
@@ -508,6 +521,45 @@ let test_attribute_validation () =
   List.iter test_good_attribute good_attributes ;
   List.iter test_bad_attribute bad_attributes
 
+let test_observed_components_of () =
+  let open Xapi_globs in
+  let open Xapi_observer_components in
+  let original_value = !observer_experimental_components in
+  let all_but_comp comp all = List.filter (fun c -> comp <> c) all in
+  let no_exp_comp = StringSet.empty in
+  let smapi_exp_comp = StringSet.singleton Constants.observer_component_smapi in
+  let observed_components_of_inputs = [[]; [SMApi]; [Xapi; SMApi]; all] in
+  let expected_outputs_no_exp_comp =
+    [all_but_comp Xapi_clusterd all; [SMApi]; [Xapi; SMApi]; all]
+  in
+  let expected_outputs_smapi_exp_comp =
+    [
+      all |> all_but_comp Xapi_clusterd |> all_but_comp SMApi
+    ; []
+    ; [Xapi]
+    ; all_but_comp SMApi all
+    ]
+  in
```

Comment:
When writing test data it's better to use tuples so the expected result and the tested data are as close as possible. On top of that, a name can be attached, which means a simple List.iter is enough to run these tests:

```
let test_data = [
    ("No experimental component is expected"
    , StringSet.empty
    , [all_but_comp Xapi_clusterd all; [SMApi]; [Xapi; SMApi]; all]
    )
  ; ("SMapi is an experimental component"
      [...]
    )
]
List.iter test_exp_comp test_data
```

Reply:
Done here e8d47962c6f3917a7235ceb069a0588d515ddfeb.